Better application paths handling in YAML testsuite

enkiusz opened the pull request:
#### Summary

Introduce a more generic and capable replacement for `ApplicationPaths`. The `ApplicationPaths` now became `SubprocInfoRepo` which is a dict-like object with a "know-how" dict and `PathsFinder` assigned to it for optional discovery of paths which were not specified. Together with some other changes this makes `run_test_suite.py` and `test_definition.py` simpler, more generic and a small bit faster.

There are now only 3 arguments used to pass paths to apps and tools to `run_test_suite.py`: `--app-path`, `--tool-path` and `--custom-path`. For example in order to point to the `chip-tool` binary now one needs to specify `--tool-path chip-tool:<path-to-chiptool>`. All of this is showcased in `tests.yaml`.

Scanning `$CHIP_ROOT` for apps which are missing is now optional and controlled by the `--discover-paths` flag.

For the purposes of smoother transition the old arguments are still working but print a deprecation warnings when they used. CI will use the new approach.

#### Testing

Local runs of the YAML testsuite.

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [X] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [X] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [X] PR size is short
-   [X] Try to avoid "squashing" and "force-update" in commit history
-   [X] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

enkiusz committed:
SubprocessInfoRepo replaces ApplicationPaths simplifying the YAML test framework and making it more generic.

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Rewrite YAML test runs in CI to use the new CLI flags for passing application and tool paths

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Adjust local.py to use new CLI path args

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Rename SubprocEntry to sound less generic, small fixes

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Remove old ApplicationPaths implementation

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Use darwin-specific tool name, it will be aliased as `chip-tool` automatically by run_test_suite.py

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Restyled fixes

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Implement improvements suggested by copilot

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Fix documentation, fix linux coverage build

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

enkiusz committed:
Allow None for PathsFinderProto

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Fix Darwin CI, improve error logging

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by autopep8

arkq commented on scripts/tests/local.py:
```diff
@@ -151,105 +154,135 @@ def _get_targets(coverage: Optional[bool]) -> list[ApplicationTarget]:
-            key="CHIP_TOOL",
+            kind=SubprocessKind.TOOL,
+            env_key="CHIP_TOOL",
+            cli_key="chip-tool",
```
What will happen if one will make a typo? E.g.:
```bash
scripts/tests/run_test_suite.py run \
    --tool-path chip-tool-with-typo:out/linux-x64-chip-tool/chip-tool
```

arkq commented on scripts/tests/chipyaml/paths_finder.py:
```diff
@@ -45,7 +45,7 @@ def get(self, target_name: str) -> str:
                 continue
 
             _PATHS_CACHE[target_name] = str(path)
```
Whey `Path` is not stored in the cache?

arkq reviewed (commented):
Is such output intentional (`None` in the message)?:
```console
$ scripts/tests/run_test_suite.py --runner chip_tool_python --log-level=debug  --dry-run run --iterations 1 --disco
ver-paths
...
Require key 'all-clusters' kind 'None' target_name 'None'
...
```

enkiusz committed:
Drop redundant ./ in paths

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Explain the difference between specifying an app and a tool in the CLI, move --dry-run option to the run command closer to where it's used

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Implement post-review improvements:
- better error messages by throwing click.BadOptionUsage where applicable
- warn against possible typos in app/tool keys
- remove redundant debug messages
- properly print SubprocessInfo when running with --dry-run

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Implement post-review improvements:

- guard BUILTIN_SUBPROC_KNOWHOW with a MappingProxyType
- put Path objects into the paths_finder.py cache instead of strings

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Fix typo in CI

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Improve exception strings, make --discover-paths work like described

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Restyle fix

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Ruff fix

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Refert --discover-paths to previous behaviour

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Fix Jupyter notebook CI correctly this time

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Merge branch 'master' into better-application-paths-v2

enkiusz commented on scripts/tests/chiptest/test_definition.py:
```diff
@@ -214,82 +216,159 @@ def __terminateProcess(self):
-                self.energy_gateway_app, self.energy_management_app, self.closure_app]
-
-    def items_with_key(self):
+class KnowhowEntry:
```
That is a better name, I will adjust.

enkiusz committed:
Implement post-review improvements:

- revert removal of redundant `./`, a new PR will be created for this
- add some explanation for the `--app-path` and `--tool-path` flags to the integration documentation
- add missing application paths to example commands
- clarify internal class and variable names in test_definition.py

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz commented on scripts/tests/run_test_suite.py:
```diff
@@ -246,58 +223,34 @@ def cmd_list(context):
-    help='what python script to use for running yaml tests using matter-repl as controller')
+    '--custom-path', multiple=True,
+    help="Set path with a custom kind, value should be <kind>:<key>:<path>, valid kind values are "
+    f"{[v.value for v in SubprocessKind.__members__.values()]}"
```
Wrong.

enkiusz committed:
Implement changes suggested by copilot:

- paths and spelling in doc
- bring back sudo removed by mistake when resolving conflicts

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Reintroduce old path options marking them as deprecated

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Shorten builtin subprocess know-how data variable name

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz commented on scripts/tests/chiptest/test_definition.py:
```diff
@@ -214,82 +216,159 @@ def __terminateProcess(self):
+    target_name: str | None = None
+
+
+BUILTIN_SUBPROC_KNOWHOW = MappingProxyType({
```
It can be `BUILTIN_SUBPROC_DATA`, fine by me.

enkiusz committed:
Merge branch 'master' into better-application-paths-v2

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

enkiusz committed:
Merge branch 'master' of https://github.com/project-chip/connectedhomeip into better-application-paths-v2

enkiusz committed:
Add missing import

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Fix issues flagged by copilot

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Fix mistake made during conflict resolution

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Merge remote-tracking branch 'upstream/master' into better-application-paths-v2

enkiusz committed:
Fix mistakes made during conflict resolution

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by isort

enkiusz committed:
Implement changes suggested by copilot

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Merge remote-tracking branch 'upstream/master' into better-application-paths-v2

enkiusz committed:
Fix mypy errors

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz commented:
@bzbarsky-apple the improvements requested have been implemented

soares-sergio commented on .github/workflows/tests.yaml:
```diff
@@ -318,22 +318,22 @@ jobs:
-                     --energy-gateway-app ./objdir-clone/linux-x64-energy-gateway-${BUILD_VARIANT}/chip-energy-gateway-app \
-                     --energy-management-app ./objdir-clone/linux-x64-energy-management-${BUILD_VARIANT}/chip-energy-management-app \
+                     --tool-path chip-tool:./objdir-clone/linux-x64-chip-tool${CHIP_TOOL_VARIANT}-${BUILD_VARIANT}/chip-tool \
+                     --app-path all-clusters:./objdir-clone/linux-x64-all-clusters-${BUILD_VARIANT}/chip-all-clusters-app \
```
I wonder if we can get rid of the explicit duplication of all these paths?

Maybe have a list of environment variables at the top and then reuse them throughout ? Something like:
```
env:
  # Define the "Dictionary" of full --app-path arguments once
  APP_PATH_ALL_CLUSTERS: "all-clusters:./objdir-clone/linux-x64-all-clusters-${BUILD_VARIANT}/chip-all-clusters-app"
  APP_PATH_LOCK: "lock:./objdir-clone/linux-x64-lock-${BUILD_VARIANT}-unified/chip-lock-app"
  APP_PATH_TV: "tv:./objdir-clone/linux-x64-tv-app-${BUILD_VARIANT}/chip-tv-app"
...


jobs:
  run_tests:
    steps:
      - name: Run Lock Test
        run: |
          ./scripts/tests/run_test_suite.py \
            --app-path ${{ env.APP_PATH_LOCK }} \
            run --target TestLockCluster
      
      - name: Run Multi-App Test
        run: |
          ./scripts/tests/run_test_suite.py \
            --app-path ${{ env.APP_PATH_ALL_CLUSTERS }} \
            --app-path ${{ env.APP_PATH_TV }} \
            run --target TestMultipleApps
```

enkiusz committed:
run_test_suite: Make an option to print help about app/tool path selection, extra trivial improvements.

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz commented on docs/testing/integration_tests.md:
```diff
@@ -47,6 +47,34 @@ way that is understandable, readable, and has the features you need
+`--discover-paths` flag can be used to automatically discover paths to
+applications and tools. The root directory of the Matter SDK is used as the
+starting path of the search.
+
```
Done.

enkiusz committed:
chipyaml: Implement post-review suggestions.

- remove the `--custom-path` option
- add infrormation on how to print known keys to `integration_tests.md`

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
docs: Fix docbuild

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
run_test_suite: Missed one custom path reference

Signed-off-by: Maciej Grela <m.grela@samsung.com>
